### PR TITLE
planner: support a hint to read from tiflash in planner

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
Support a hint to read from tiflash in planner.

### What is changed and how it works?
It's almost same as other hints. But the tests are changed much.

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to update the documentation

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -377,6 +379,22 @@ func (p *LogicalJoin) setPreferredJoinType(hintInfo *tableHintInfo) {
...
+func (ds *DataSource) setPreferredStoreType(hintInfo *tableHintInfo) {
+	if hintInfo == nil {
+		return
+	}
+
+	var alias *hintTableInfo
+	if len(ds.TableAsName.L) != 0 {
+		alias = &hintTableInfo{name: *ds.TableAsName, selectOffset: ds.SelectBlockOffset()}
+	} else {
+		alias = extractTableAlias(ds)
```
Comment (winoros) [inline planner/core/logical_plan_builder.go]: You can directly use the datasource's original table name. Don't need to call this function.

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -2009,16 +2028,21 @@ func (b *PlanBuilder) pushTableHints(hints []*ast.TableOptimizerHint, nodeType n
 					},
 				})
 			}
+		case HintReadFromStorage:
+			if hint.StoreType.L == "tiflash" {
```
Comment (winoros) [inline planner/core/logical_plan_builder.go]: Can we mark this as a const value?

File: planner/core/logical_plans.go [hunk 3]
```diff
@@ -102,6 +102,8 @@ const (
 	preferMergeJoin
 	preferHashAgg
 	preferStreamAgg
+	preferTiKV
```
Comment (winoros) [inline planner/core/logical_plans.go]: I vote for that putting this two as single const block.

File: planner/core/find_best_task.go [hunk 4]
```diff
@@ -1014,7 +1019,13 @@ func (ds *DataSource) getOriginalPhysicalTableScan(prop *property.PhysicalProper
 		Ranges:          path.ranges,
 		AccessCondition: path.accessConds,
 		filterCondition: path.tableFilters,
+		StoreType:       kv.TiKV,
```
Comment (foreyes) [inline planner/core/find_best_task.go]: Can we remove this line? Since `StoreType` will be set below.

File: planner/core/common_plans.go [hunk 5]
```diff
@@ -698,22 +699,31 @@ func (e *Explain) explainPlanInRowFormat(p Plan, taskType, indent string, isLast
 
 	switch x := p.(type) {
 	case *PhysicalTableReader:
-		e.explainPlanInRowFormat(x.tablePlan, "cop", childIndent, true)
+		var storeType string
+		switch x.StoreType {
+		case kv.TiKV:
+			storeType = HintTiKV
```
Comment (foreyes) [inline planner/core/common_plans.go]: I think using `HintTiKV` here makes it a little bit confuse.

File: planner/core/task.go [hunk 6]
```diff
@@ -968,7 +985,14 @@ func (p *PhysicalStreamAgg) attach2Task(tasks ...task) task {
 	t := tasks[0].copy()
 	inputRows := t.count()
 	if cop, ok := t.(*copTask); ok {
-		partialAgg, finalAgg := p.newPartialAggregate()
+		// copToFlash means whether the cop task is run on flash storage
+		copToFlash := false
+		if cop.tablePlan != nil {
+			if ts, ok := cop.tablePlan.(*PhysicalTableScan); ok {
```
Comment (lamxTyler) [inline planner/core/task.go]: Is it possible the the child is `selection`?

Comment (lzmhhh123) [reply planner/core/task.go]: Oh. It's possible.

```sql
tidb(localhost:4000) > desc select count(b) from t1 where t1.b=1 group by a;
+--------------------------+----------+-----------+-------------------------------------------------------------+
| id                       | count    | task      | operator info                                               |
+--------------------------+----------+-----------+-------------------------------------------------------------+
| HashAgg_13               | 8.00     | root      | group by:Column#7, funcs:count(Column#6)                    |
| └─TableReader_14         | 8.00     | root      | data:HashAgg_6                                              |
|   └─HashAgg_6            | 8.00     | cop[tikv] | group by:Column#1, funcs:count(Column#2)                    |
|     └─Selection_12       | 10.00    | cop[tikv] | eq(Column#2, 1)                                             |
|       └─TableScan_11     | 10000.00 | cop[tikv] | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo |
+--------------------------+----------+-----------+-------------------------------------------------------------+
5 rows in set (0.00 sec)
```

File: planner/core/exhaust_physical_plans.go [hunk 7]
```diff
@@ -464,7 +464,7 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *property.PhysicalProperty, ou
 	}
 	ds, isDataSource := innerChild.(*DataSource)
 	us, isUnionScan := innerChild.(*LogicalUnionScan)
-	if !isDataSource && !isUnionScan {
+	if (!isDataSource && !isUnionScan) || (isDataSource && ds.preferStoreType&preferTiFlash != 0) {
```
[the remark this answers is not in the document]
Comment (lzmhhh123) [reply planner/core/exhaust_physical_plans.go]: Done.

File: planner/core/common_plans.go [hunk 8]
```diff
@@ -698,22 +699,31 @@ func (e *Explain) explainPlanInRowFormat(p Plan, taskType, indent string, isLast
...
 	switch x := p.(type) {
 	case *PhysicalTableReader:
-		e.explainPlanInRowFormat(x.tablePlan, "cop", childIndent, true)
+		var storeType string
+		switch x.StoreType {
+		case kv.TiKV:
+			storeType = "tikv"
+		case kv.TiFlash:
+			storeType = "tiflash"
+		default:
```
[the remark this answers is not in the document]
Comment (lzmhhh123) [reply planner/core/common_plans.go]: Good idea.

[the remark this answers is not in the document]
Comment (lzmhhh123) [reply planner/core/common_plans.go]: Done.

**Pull request merged**